File: src/main/java/org/styl/gravitus/engine/UniverseEngine.java
package org.styl.gravitus.engine;

import java.util.List;

import org.styl.gravitus.Specs;
import org.styl.gravitus.entities.SpaceObject;

import lombok.Getter;
import lombok.Setter;

public class UniverseEngine {

	@Getter @Setter private List<SpaceObject> objects;

	public void estimateTick() {
		objects.forEach(o -> calculate(o));
		objects.forEach(o -> update(o));
	}

	private void calculate(SpaceObject o) {
		objects.forEach((a) -> {
			if (o != a) {
				long dist = distance(o, a);
				double angle = angle(o, a);
				double g = Specs.instance.G * a.getMass() / Math.pow(dist, 2);
				double gx = g * Math.cos(angle);
				double gy = g * Math.sin(angle);
				o.incrAccx(gx);
				o.incrAccy(gy);
			}
		});
	}

	private void update(SpaceObject o) {
		o.incrVelx(o.getAccx() * Clock.INSTANCE.t());
		o.incrVely(o.getAccy() * Clock.INSTANCE.t());
		o.incrPosx(o.getVelx() * Clock.INSTANCE.t());
		o.incrPosy(o.getVely() * Clock.INSTANCE.t());
		o.setAccx(0);
		o.setAccy(0);
	}

	private long distance(SpaceObject o, SpaceObject a) {
		long dx = a.getPosx() - o.getPosx();
		long dy = a.getPosy() - o.getPosy();
		long sumsqr = (long) (Math.pow(dx, 2) + Math.pow(dy, 2));
		return (long) Math.sqrt(sumsqr);
	}

	private double angle(SpaceObject o, SpaceObject a) {
		long dx = a.getPosx() - o.getPosx();
		long dy = a.getPosy() - o.getPosy();
		return Math.atan2(dy, dx);
	}

}

File: src/main/java/org/styl/gravitus/engine/UniverseEngine.kt
package org.styl.gravitus.engine

import org.styl.gravitus.Specs
import org.styl.gravitus.entities.SpaceObject

class UniverseEngine {

	var objects: MutableList<SpaceObject> = mutableListOf()

	fun estimateTick() {
		objects.forEach { calculate(it) }
		objects.forEach { update(it) }
	}

	private fun calculate(o: SpaceObject) {
		objects.forEach { a ->
			if (o != a) {
				val dist = distance(o, a)
				val angle = angle(o, a)
				val g = Specs.instance.G * a.mass / Math.pow(dist.toDouble(), 2.0)
				val gx = g * Math.cos(angle)
				val gy = g * Math.sin(angle)
				o.acceleration.x += gx
				o.acceleration.y += gy
			}
		}
	}

	private fun update(o: SpaceObject) {
		val t = Clock.INSTANCE.t().toDouble()
		o.velocity.x += o.acceleration.x * t
		o.velocity.y += o.acceleration.y * t
		o.position.x += o.velocity.x * t
		o.position.y += o.velocity.y * t
		o.acceleration.x = 0.0
		o.acceleration.y = 0.0
	}

	private fun distance(o: SpaceObject, a: SpaceObject): Long {
		val dx = a.position.x - o.position.x
		val dy = a.position.y - o.position.y
		val sumsqr = dx * dx + dy * dy
		return Math.sqrt(sumsqr).toLong()
	}

	private fun angle(o: SpaceObject, a: SpaceObject): Double {
		val dx = a.position.x - o.position.x
		val dy = a.position.y - o.position.y
		return Math.atan2(dy, dx)
	}
}
